### PR TITLE
Query both DPA and LPI datasets on OS PLaces API calls

### PR DESCRIPTION
**Pull request by catalinailie**
We need to change our calls to OS Places API so we look in both DPA and LPI datasets.

Trello card: https://trello.com/c/156rUuMP/2925-look-at-multiple-database-config-for-calls-to-os-places-api-2
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

### CODE REVIEW

File: spec/lib/os_places_api/client_spec.rb [hunk 1]
```diff
@@ -197,6 +197,27 @@
...
     end
 
+    context "the postcode doesn't exist in the DPA dataset" do
+      before :each do
+        Postcode.where(postcode: postcode).map(&:destroy)
+      end
+
+      it "should query OS Places API and return results from LPI dataset" do
+        os_places_api_results[0]["LPI"] = os_places_api_results[0]["DPA"]
+        os_places_api_results[0].delete("DPA")
```
Comment (ChrisBAshton) [inline spec/lib/os_places_api/client_spec.rb:207]: I don't think it's safe to reuse the DPA result, as the LPI result is in a different format.

Here's a slimmed down extract from https://api.os.uk/search/places/v1/postcode?key=API_KEY_GOES_HERE&postcode=E18QS&dataset=DPA,LPI&output_srs=WGS84:

```json
{
    "results": [
        {
            "DPA": {
                "UPRN": "6714279",
                "ADDRESS": "1, WHITECHAPEL HIGH STREET, LONDON, E1 8QS",
                "POSTCODE": "E1 8QS",
                "LNG": -0.0729707,
                "LAT": 51.5144426,
                ...
            }
        },
        {
            "LPI": {
                "UPRN": "6714279",
                "ADDRESS": "1, WHITECHAPEL HIGH STREET, LONDON, TOWER HAMLETS, E1 8QS",
                "LNG": -0.0729707,
                "LAT": 51.5144426,
                ...
            }
        }
    ]
}
```

The LPI result has different keys to the DPA one, but the only one that is a concern is the lack of `POSTCODE` key, which could be problematic [here](https://github.com/alphagov/locations-api/blob/eb60df6cdd397bf97e8f83ba5717ec29063b510a/lib/os_places_api/location_builder.rb#L8).

Could you add an LPI result to the `os_places_api_results` variable at the top of the file (even though it has a lot of duplication to the DPA result)?

File: lib/os_places_api/location_builder.rb [hunk 2]
```diff
@@ -4,12 +4,12 @@ def initialize(result)
...
     end
 
-    def build_location
-      Location.new(postcode: @result.dig("DPA", "POSTCODE"),
-                   address: @result.dig("DPA", "ADDRESS"),
-                   latitude: @result.dig("DPA", "LAT"),
-                   longitude: @result.dig("DPA", "LNG"),
-                   local_custodian_code: @result.dig("DPA", "LOCAL_CUSTODIAN_CODE"))
+    def build_location(dataset)
+      Location.new(postcode: @result.dig(dataset, "POSTCODE"),
```
Comment (ChrisBAshton) [inline lib/os_places_api/location_builder.rb:8]: Looking again, I think we can actually remove this parameter - no need for us to store the `postcode` in the JSON when we already have a sanitised & normalised `postcode` field in the same row in the database. That will also require a tweak to the Pact tests though 😅 

It feels like a separate piece of work, except that we're only having to do this work because of the missing `POSTCODE` field in the LPI results, so also fine for this work to be bundled into this PR.

**Review by ChrisBAshton — COMMENTED** [review]
Ah, nice catch! ⭐ 

I've made some suggestions to make this a bit more robust to future changes to the API.

File: lib/os_places_api/client.rb [hunk 3]
```diff
@@ -47,7 +47,13 @@ def validate_response_code(response)
...
     def build_locations(results)
       locations = results.map do |result|
-        LocationBuilder.new(result).build_location
+        LocationBuilder.new(result).build_location("DPA")
+      end
+      # if the postcode has no information in DPA dataset, use the results from LPI dataset
+      if locations[0].latitude.nil?
+        locations = results.map do |result|
+          LocationBuilder.new(result).build_location("LPI")
+        end
```
[the remark this answers is not in the document]
Comment (catalinailie) [reply lib/os_places_api/client.rb:56]: I just checked and yes, we can use UPRN in this case. This is a way nicer solution 🙌

File: lib/os_places_api/location_builder.rb (hunk 2 above)
Comment (catalinailie) [reply lib/os_places_api/location_builder.rb:8]: The OS API response is stored raw in the db. The Location instance is used for building informations about a specific postcode. But I agree we don't need the postcode from the api response as long as we have the sanitised one. Also, `Locations.postcode` is not used anywhere yet so we can remove it.

Comment (ChrisBAshton) [reply lib/os_places_api/location_builder.rb]: Ah yes, I misremembered!

File: lib/os_places_api/client.rb [hunk 4]
```diff
@@ -46,8 +46,14 @@ def validate_response_code(response)
...
-      locations = results.map do |result|
-        LocationBuilder.new(result).build_location
+      uprns_used = []
+      locations = []
+      results.each do |result|
+        result_hash = result[result.keys.first] # it is irrelevant if it's LPI or DPA
+        unless uprns_used.include?(result_hash["UPRN"])
+          uprns_used << result_hash["UPRN"]
+          locations << LocationBuilder.new(result).build_location(result.keys.first)
+        end
```
[the remark this answers is not in the document]
Comment (catalinailie) [reply lib/os_places_api/client.rb]: Yes, this is super clean 🥇

File: lib/os_places_api/location_builder.rb [hunk 5]
```diff
@@ -4,12 +4,11 @@ def initialize(result)
       @result = result
     end
 
-    def build_location
-      Location.new(postcode: @result.dig("DPA", "POSTCODE"),
-                   address: @result.dig("DPA", "ADDRESS"),
-                   latitude: @result.dig("DPA", "LAT"),
-                   longitude: @result.dig("DPA", "LNG"),
-                   local_custodian_code: @result.dig("DPA", "LOCAL_CUSTODIAN_CODE"))
+    def build_location(dataset)
```
Comment (ChrisBAshton) [inline lib/os_places_api/location_builder.rb]: No need to add this parameter - looks like this slipped in from the next commit 👀 

Also this first commit has a typo ("form" instead of "from")

File: spec/lib/os_places_api/client_spec.rb [hunk 6]
```diff
@@ -196,6 +239,26 @@
...
+          .to_return(status: 200, body: successful_response.to_json)
+
+        expect(client.locations_for_postcode(postcode).as_json).to eq(
+          {
+            "average_latitude" => average_latitude,
+            "average_longitude" => average_longitude,
+            "results" => [location].as_json,
+          },
+        )
+      end
```
Comment (ChrisBAshton) [inline spec/lib/os_places_api/client_spec.rb:259]: I imagine if you delete line 248 this test still passes.
Could we add an explicit test that ensures that duplicate `UPRN` results are filtered out?
And also a test that both LPI and DPA results are returned if they have different UPRNs?

**Review by ChrisBAshton — COMMENTED** [review]
This is looking great! Close to merge now 🔧

File: spec/lib/os_places_api/client_spec.rb [hunk 7]
```diff
@@ -85,7 +134,21 @@
           {
             "average_latitude" => average_latitude,
             "average_longitude" => average_longitude,
-            "results" => [location].as_json,
+            "results" => locations.as_json,
+          },
+        )
+      end
+
+      it "should query OS Places API and return unique results" do
```
Comment (ChrisBAshton) [inline spec/lib/os_places_api/client_spec.rb:142]: Nitpick - this reads more clearly to me:

```suggestion
      it "should filter out duplicate `UPRN` results from the OS Places API response" do
```

**Review by ChrisBAshton — APPROVED** [review]
Great stuff! ⭐
